pyulog/db.py

```diff
@@ -0,0 +1,572 @@
+                        timestamp=timestamp,
+                        message=message,
+                    )
+                )
+
+            # message_formats
+            cur.execute('''
+                SELECT
+                    msg.Name,
+                    field.FieldType,
+                    field.ArraySize,
+                    field.Name
+                FROM ULogMessageFormat msg JOIN ULogMessageFormatField field
+                    ON field.MessageId = msg.Id
+                WHERE ULogId = ?
+                ''', (self._pk,))
+            for row in cur.fetchall():
+                msg_name = row[0]
+                field_data = row[1:]
+                if msg_name in self._message_formats:
+                    self._message_formats[msg_name].fields.append(field_data)
+                else:
+                    self._message_formats[msg_name] = DatabaseULog.DatabaseMessageFormat(
+                        name=msg_name,
+                        fields=[field_data]
+                    )
+
+            # msg_info_dict
+            cur.execute('''
+                SELECT Key, Value
+                FROM ULogMessageInfo
+                WHERE ULogId = ?
+                ''', (self._pk,))
+            for key, value in cur.fetchall():
+                self._msg_info_dict[key] = value
+
+            # msg_info_multiple_dict
+            cur.execute('''
+                SELECT Id, Key
+                FROM ULogMessageInfoMultiple
+                WHERE ULogId = ?
+                ''', (self._pk,))
+            for message_id, key in cur.fetchall():
+                self._msg_info_multiple_dict[key] = []
+                cur.execute('''
+                    SELECT Id
+                    FROM ULogMessageInfoMultipleList
+                    WHERE MessageId = ?
+                    ORDER BY SeriesIndex
+                    ''', (message_id,))
+                for (list_id,) in cur.fetchall():
+                    cur.execute('''
+                        SELECT Value
+                        FROM ULogMessageInfoMultipleListElement
+                        WHERE ListId = ?
+                        ORDER BY SeriesIndex
+                        ''', (list_id,))
+                    self._msg_info_multiple_dict[key].append([value for (value,) in cur.fetchall()])
+
+            # initial_parameters
+            cur.execute('''
+                SELECT Key, Value
+                FROM ULogInitialParameter
+                WHERE ULogId = ?
+                ''', (self._pk,))
+            for key, value in cur.fetchall():
+                self._initial_parameters[key] = value
+
+            cur.close()
+
+    def get_dataset(self, name, multi_instance=0):
+        '''
+        Access a specific dataset and its data series from the database.
+
+        Note that the parameters name and multi_instance parameters are named
+        the same way as the original ULog file for compatibility.
+        '''
+        with self._db() as con:
```

Comment:
Is opening the db connection cheap?

Reply:
In general no, but I'm not sure if I understand the intent behind the question.

Reply:
Depending on the use-case, this method might be called a lot, so I'd make sure to keep the connection open.

Reply:
Storing an open connection would require always opening/closing when using DatabaseULog, which is pretty cumbersome. Instead I added an option of supplying a cursor directly if one is doing a lot of get_dataset. Updated PR coming soon.